tensorflow/core/kernels/training_ops.cc

```diff
@@ -310,21 +310,63 @@ struct ApplyAdamNonCuda {
                   typename TTypes<T>::ConstScalar beta2,
                   typename TTypes<T>::ConstScalar epsilon,
                   typename TTypes<T>::ConstFlat grad, bool use_nesterov) {
+    const int length = var.size();
+
+    T* var_ptr = var.data();
+    T* m_ptr = m.data();
+    T* v_ptr = v.data();
+    const T* g_ptr = grad.data();
     const T alpha = lr() * Eigen::numext::sqrt(T(1) - beta2_power()) /
                     (T(1) - beta1_power());
     // beta1 == μ
     // beta2 == ν
     // v     == n
     // var   == θ
 
-    m.device(d) += (grad - m) * (T(1) - beta1());
-    v.device(d) += (grad.square() - v) * (T(1) - beta2());
-    if (use_nesterov) {
-      var.device(d) -= ((grad * (T(1) - beta1()) + beta1() * m) * alpha) /
-                       (v.sqrt() + epsilon());
-    } else {
-      var.device(d) -= (m * alpha) / (v.sqrt() + epsilon());
-    }
+    auto shard = [this, var_ptr, m_ptr, v_ptr, g_ptr, alpha, beta1, beta2,
+                  epsilon, use_nesterov](int begin, int end) {
+      T m;
+      T v;
+      T g;
+
+      if (use_nesterov) {
+        for (int i = begin; i < end; ++i) {
+          m = *(m_ptr + i);
+          v = *(v_ptr + i);
+          g = *(g_ptr + i);
+
+          m += (g - m) * (T(1) - beta1());
+          v += (g * g - v) * (T(1) - beta2());
+          *(var_ptr + i) -= ((g * (T(1) - beta1()) + beta1() * m) * alpha) /
+                            (Eigen::numext::sqrt(v) + epsilon());
+          *(m_ptr + i) = m;
+          *(v_ptr + i) = v;
+        }
+      } else {
+        for (int i = begin; i < end; ++i) {
+          m = *(m_ptr + i);
+          v = *(v_ptr + i);
+          g = *(g_ptr + i);
+
+          m += (g - m) * (T(1) - beta1());
+          v += (g * g - v) * (T(1) - beta2());
+          *(var_ptr + i) -= (m * alpha) / (Eigen::numext::sqrt(v) + epsilon());
+          *(m_ptr + i) = m;
+          *(v_ptr + i) = v;
+        }
+      }
+    };
+
+    // TODO: lack of memory access and sqrt
+    int cost = Eigen::TensorOpCost::AddCost<int>() * 6 +
+               Eigen::TensorOpCost::AddCost<T>() * 4 +
+               // Consider Sub as Add
+               Eigen::TensorOpCost::AddCost<T>() * 5 +
+               Eigen::TensorOpCost::MulCost<T>() * 6 +
+               Eigen::TensorOpCost::DivCost<T>();
+    auto worker_threads = *(ctx->device()->tensorflow_cpu_worker_threads());
+    Shard(worker_threads.num_threads, worker_threads.workers, length, cost,
```

Comment:
Please add a comment that because of memory access pattern/locality we are using Shard instead of "regular" tensor expressions.

Reply:
Done.